Add coordinator to 17Track

shaiu opened the pull request:
## Proposed change
Add coordinator to 17Track

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure


[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

shaiu committed:
Add coordinator to 17Track

gjohansson-ST commented:
There are more in here than just adding a DUC. So I would suggest you break out the refactoring and entity descriptions into their own PR's to make it more feasible to review. Thanks

shaiu committed:
Add coordinator to 17Track

remove SensorEntityDescription (different PR)

shaiu committed:
Merge branch 'dev' into shaiu-add-coordinator

shaiu commented:
@gjohansson-ST I removed the SensorEntityDescription
I'm not sure how I can remove more than that in this PR.
This code is a bit complicated as it has "dynamic" sensors added and removed
I plan (together with the help of @joostlek) to refactor this integration to have service calls instead of the "dynamic" sensors in different PRs

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,137 @@
+    LOGGER,
+)
+
+
+class SeventeenTrackData:
+    """Class for handling the data retrieval."""
+
+    def __init__(self) -> None:
+        """Initialize the data object."""
+        self.summary: dict[str, dict[str, Any]] = {}
+        self.current_packages: dict[str, dict[str, Any]] = {}
+        self.new_packages: dict[str, Package] = {}
+        self.old_packages: set[Package] = set()
```
This can be a dataclass

shaiu replied:
Added

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,137 @@
+        except SeventeenTrackError as err:
+            LOGGER.error("There was an error retrieving the summary: %s", err)
+
+        current_packages = set(
+            await self._client.profile.packages(show_archived=self._show_archived)
+        )
```
Can this also raise?

shaiu replied:
Yes.  Added try catch

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,137 @@
+            summary = await self._client.profile.summary(
+                show_archived=self._show_archived
+            )
+            for status, quantity in summary.items():
+                data.summary[slugify(status)] = {
+                    "quantity": quantity,
+                    "packages": [],
+                }
+
```
Keep things that can't raise out of the try block

shaiu replied:
moved

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +94,160 @@ async def async_setup_entry(
         self._status = status
-        self._attr_name = f"Seventeentrack Packages {status}"
-        self._attr_unique_id = f"summary_{data.account_id}_{slugify(status)}"
+        self._attr_extra_state_attributes = {}
```
Unneeded

shaiu replied:
removed

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +94,160 @@ async def async_setup_entry(
+    ) -> None:
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        self._attr_extra_state_attributes = {}
```
Idem

shaiu replied:
removed

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +94,160 @@ async def async_setup_entry(
     def available(self) -> bool:
         """Return whether the entity is available."""
-        return self._data.packages.get(self._tracking_number) is not None
+        package_data = self.coordinator.data.current_packages.get(
+            self._tracking_number, {}
+        )
+        return package_data.get("package") is not None
```
```suggestion
        return self._tracking_number in self.coordinator.data.current_packages
```

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +94,160 @@ async def async_setup_entry(
     def name(self) -> str:
         """Return the name."""
-        if not (name := self._friendly_name):
+        package_data = self.coordinator.data.current_packages.get(
+            self._tracking_number, {}
+        )
+        package = package_data.get("package")
+        if package is None or not (name := package.friendly_name):
```
Afaik (I know for sure for native value and some other properties, name I am not sure), if the available property returns False, the other properties aren't requested anymore, so we shouldn't have to check again here.

(We can use square brackets instead of get

self.coordinator.data.current_packages[self._tracking_number]
)

shaiu replied:
moved it anyway to init

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +94,160 @@ async def async_setup_entry(
-
-        identification = (
-            self._friendly_name if self._friendly_name else self._tracking_number
+        package_data = self.coordinator.data.current_packages.get(
+            self._tracking_number, {}
         )
```
Idem

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +94,160 @@ async def async_setup_entry(
 
-class SeventeenTrackData:
-    """Define a data handler for 17track.net."""
+def remove_entity(hass: HomeAssistant, account_id: str, tracking_number: str) -> bool:
```
Can we maybe just pass the list of packages it should remove so it can do a for loop here?

shaiu replied:
done

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +94,160 @@ async def async_setup_entry(
-                )
+def notify_delivered(hass: HomeAssistant, friendly_name: str, tracking_number: str):
+    """Notify when package is delivered."""
+    LOGGER.info("Package delivered: %s", tracking_number)
```
Please don't use the info level of logging

shaiu replied:
changed to debug

shaiu committed:
Update homeassistant/components/seventeentrack/coordinator.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Update homeassistant/components/seventeentrack/sensor.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Add coordinator to 17Track

fix CR

shaiu committed:
Merge branch 'dev' into shaiu-add-coordinator

shaiu committed:
Merge branch 'dev' into shaiu-add-coordinator

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,143 @@
+class SeventeenTrackData:
+    """Class for handling the data retrieval."""
+
+    def __init__(self) -> None:
+        """Initialize the data object."""
+        self.summary: dict[str, dict[str, Any]] = {}
+        self.current_packages: dict[str, dict[str, Any]] = {}
+        self.new_packages: dict[str, Package] = {}
+        self.old_packages: set[Package] = set()
```
```suggestion
    summary: dict[str, dict[str, Any]]
    current_packages: dict[str, dict[str, Any]]
```

I think it's wise to actually just keep the list of package tracking number of the last update in a set in `async_setup_entry` in `sensor.py`. This would make the coordinator less responsible for creating and removing entities. (and it concentrates the logic that we can remove in a few months)

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,143 @@
+        self._packages = (self._packages | packages_to_add) - packages_to_remove
+
+        for package in current_packages:
+            data.current_packages[package.tracking_number] = {
+                "package": package,
+                "extra": {
+                    ATTR_DESTINATION_COUNTRY: package.destination_country,
+                    ATTR_INFO_TEXT: package.info_text,
+                    ATTR_TIMESTAMP: package.timestamp,
+                    ATTR_LOCATION: package.location,
+                    ATTR_ORIGIN_COUNTRY: package.origin_country,
+                    ATTR_PACKAGE_TYPE: package.package_type,
+                    ATTR_TRACKING_INFO_LANGUAGE: package.tracking_info_language,
+                    ATTR_TRACKING_NUMBER: package.tracking_number,
+                },
+            }
```
If I recall correctly, these were the extra state attributes. I would keep that logic at the sensor.

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,143 @@
+
+        return data
+
+    @property
+    def show_delivered(self):
+        """Return whether delivered packages should be shown."""
+        return self._show_delivered
+
+    @property
+    def account_id(self):
+        """Return the account ID."""
+        return self._account_id
```
instead of returning the `self._something` just assign the variable to `self.something` in the first place so you dont need to create separate properties

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +95,150 @@ async def async_setup_entry(
+            SeventeenTrackPackageSensor(
+                coordinator, t_number, p_data.friendly_name, p_data.status
+            )
+            for t_number, p_data in coordinator.data.new_packages.items()
```
I'd write out the variables, we write code for developers, not for the computer :)

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +95,150 @@ async def async_setup_entry(
         self._status = status
-        self._attr_name = f"Seventeentrack Packages {status}"
-        self._attr_unique_id = f"summary_{data.account_id}_{slugify(status)}"
+        self._attr_name = f"Seventeentrack Packages {self._status}"
```
```suggestion
        self._attr_name = f"Seventeentrack Packages {status}"
```

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +95,150 @@ async def async_setup_entry(
     def available(self) -> bool:
         """Return whether the entity is available."""
-        return self._state is not None
+        return self.coordinator.data.summary[self._status]["quantity"] is not None
```
When this is unavalable, is summary[status] even set? Shouldn't it be

```suggestion
        return self._status in self.coordinator.data.summary
```

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +95,150 @@ async def async_setup_entry(
-            data.account_id, self._tracking_number
+            coordinator.account_id, tracking_number
         )
+        name = friendly_name if friendly_name else tracking_number
+        self._name = f"Seventeentrack Package: {name}"
```
Why did we move this again?

shaiu replied:
there is no need to calculate this each time property name is called.
We have the information at __init__ time

shaiu committed:
Add coordinator to 17Track

fix second CR

shaiu committed:
Add coordinator to 17Track

remove commented out code + fix display name

shaiu committed:
Merge branch 'dev' into shaiu-add-coordinator

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,93 @@
+        )
+        self.show_delivered = self.config_entry.options[CONF_SHOW_DELIVERED]
+        self.account_id = client.profile.account_id
+        self.data = SeventeenTrackData()
```
Why don't you construct a new one every time?

shaiu replied:
fixed

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +108,193 @@ async def async_setup_entry(
 
-    client = hass.data[DOMAIN][config_entry.entry_id]
+    coordinator: SeventeenTrackCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    @callback
+    def _async_create_remove_entities():
+        old_packages = coordinator.data.current_packages - set(
+            coordinator.data.live_packages.values()
+        )
+        packages_to_add = (
+            set(coordinator.data.live_packages.values())
+            - coordinator.data.current_packages
+        )
```
You can create a set outside of this function with all the tracking numbers for which you already made sensors, that would simplify the coordinator and this function

shaiu replied:
cool. Done

shaiu committed:
Add coordinator to 17Track

created a set outside _async_create_remove_entities function

shaiu committed:
Merge remote-tracking branch 'origin/shaiu-add-coordinator' into shaiu-add-coordinator

shaiu committed:
Merge branch 'dev' into shaiu-add-coordinator

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,92 @@
+        """Fetch data from 17Track API."""
+
+        data = SeventeenTrackData()
+        summary = {}
+        live_packages = set()
```
Can be removed now

shaiu replied:
why? I need it when getting the packages from the API.
Then I populate a dict [str, Package] to return as the coordinator data

shaiu replied:
removed

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,92 @@
+        except SeventeenTrackError as err:
+            LOGGER.error("There was an error retrieving the packages: %s", err)
+
+        data.live_packages.clear()
```
Can also be removed

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,92 @@
+            if summary_value:
+                summary_value["packages"].append(package)
+
+        return data
```
I would almost say like, just make it like you return

`return SeventeenTrackData(packages=...)`

shaiu replied:
Got it.
Done

joostlek commented on homeassistant/components/seventeentrack/sensor.py:
```diff
@@ -111,242 +108,187 @@ async def async_setup_entry(
 
-    client = hass.data[DOMAIN][config_entry.entry_id]
+    coordinator: SeventeenTrackCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+    current_packages: set[Package] = set()
```
Since the content of the set is unique, i am not sure if package works as type. If we can prove that this works, great, otherwise i think tracking numbers are the way to go

shaiu replied:
Package holds the tracking number so it's unique. Also I need the Package itself for the various data it holds

joostlek replied:
I know it's unique for us, but I'm talking about unique in the set. If a package is set to "Delivered", is that counted as a different object? If so, this code would not work

joostlek replied:
Checkout https://github.com/home-assistant/core/blob/988c71ecc533238a1b4e82c2d9f8ec95dfd76fe0/homeassistant/components/opensky/coordinator.py#L85-L90

This is what I use in Opensky to track the flights.

shaiu replied:
Done

joostlek reviewed (changes requested):
I'll look more thorough tomorrow

shaiu committed:
Add coordinator to 17Track

fix CR

shaiu committed:
Merge remote-tracking branch 'origin/shaiu-add-coordinator' into shaiu-add-coordinator

shaiu committed:
Merge branch 'dev' into shaiu-add-coordinator

shaiu committed:
Add coordinator to 17Track

fix CR 2

shaiu committed:
Merge branch 'dev' into shaiu-add-coordinator

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,91 @@
+        self._show_archived = self.config_entry.options[CONF_SHOW_ARCHIVED]
+        self._client = client
+
+    async def _async_update_data(
+        self,
+    ) -> SeventeenTrackData:
```
```suggestion
    async def _async_update_data(self) -> SeventeenTrackData:
```

joostlek commented on homeassistant/components/seventeentrack/coordinator.py:
```diff
@@ -0,0 +1,91 @@
+        live_packages = set()
+        live_packages_dict = {}
+
+        try:
+            summary = await self._client.profile.summary(
+                show_archived=self._show_archived
+            )
+
+        except SeventeenTrackError as err:
+            LOGGER.error("There was an error retrieving the summary: %s", err)
+
+        for status, quantity in summary.items():
+            summary_dict[slugify(status)] = {
+                "quantity": quantity,
+                "packages": [],
+                "status_name": status,
+            }
+
+        try:
+            live_packages = set(
+                await self._client.profile.packages(show_archived=self._show_archived)
+            )
+        except SeventeenTrackError as err:
+            LOGGER.error("There was an error retrieving the packages: %s", err)
```
Should we raise `UpdateFailed` when one of these 2 calls fail?

shaiu replied:
I'm not sure what is the added value for raising it. If either has an error we can continue with empty data for one but the other will have data populated

joostlek replied:
How stable is the API? If its not excessively flakey I would suggest just raising it

shaiu replied:
OK. raising

joostlek replied:
These 2 calls can now be merged

shaiu replied:
merged. Thanks

shaiu committed:
Update homeassistant/components/seventeentrack/coordinator.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

shaiu committed:
Add coordinator to 17Track

raise UpdateFailed if API throws an exception

shaiu committed:
Merge branch 'dev' into shaiu-add-coordinator

shaiu committed:
Add coordinator to 17Track

merge calls

shaiu committed:
Merge branch 'dev' into shaiu-add-coordinator
